code/modules/mob/living/carbon/human/species_types/jellypeople.dm

```diff
@@ -479,88 +479,263 @@
+ */
+/datum/species/oozeling/stargazer/proc/validate_mind(datum/mind/target_mind, initial_connection = FALSE, silent = FALSE)
+	. = TRUE
+	if(QDELETED(target_mind?.current) || target_mind.current.stat == DEAD)
 		return FALSE
-	if(HAS_TRAIT(M, TRAIT_MINDSHIELD)) //mindshield implant, no dice
+	if(initial_connection && GLOB.slime_links_by_mind[target_mind])
+		var/datum/weakref/other_link_ref = GLOB.slime_links_by_mind[target_mind]
+		var/datum/species/oozeling/stargazer/other_link = other_link_ref?.resolve()
+		// If they're already slime linked, then we can't link to them.
+		if(other_link && istype(other_link))
+			return FALSE
+	if(istype(target_mind.current.get_item_by_slot(ITEM_SLOT_HEAD), /obj/item/clothing/head/foilhat))
+		if(!silent)
+			var/mob/living/carbon/human/owner = slimelink_owner.resolve()
+			if(owner)
+				to_chat(target_mind.current, "<span class='danger'>[owner]'s no-good syndicate mind-slime is blocked by your protective headgear!</span>")
+		return FALSE
+	if(HAS_TRAIT_NOT_FROM(target_mind.current, TRAIT_MINDSHIELD, "nanites")) //mindshield implant, no dice
 		return FALSE
-	var/mob/living/carbon/human/owner = slimelink_owner.resolve()
-	if(istype(M.get_item_by_slot(ITEM_SLOT_HEAD), /obj/item/clothing/head/foilhat))
-		if(owner)
-			to_chat(M, "<span class='danger'>[owner.real_name]'s no-good syndicate mind-slime is blocked by your protective headgear!</span>")
 
+/**
+ * Handles the death of a linked mob, unlinking them from the slime link.
+ *
+ * Arguments
+ * * source_mob: The mob that died.
+ */
+/datum/species/oozeling/stargazer/proc/link_death(mob/living/source_mob)
+	SIGNAL_HANDLER
+	if(!source_mob.mind || !(source_mob.mind in linked_minds))
+		return
+	var/datum/weakref/owner_ref = GLOB.slime_links_by_mind[source_mob.mind]
+	var/datum/species/oozeling/stargazer/stargazer = owner_ref?.resolve()
+	var/mob/living/carbon/human/link_owner = stargazer?.slimelink_owner?.resolve()
+	if(!link_owner || !istype(link_owner))
+		return
+	to_chat("<span class='danger'>As you die, you feel your link to [link_owner.real_name] fizzle out!</span>")
+	unlink_mind(source_mob.mind)
+
+/**
+ * Links the mind of another mob to this stargazer's slime link.
+ *
+ * Arguments
+ * * target_mind: The target mind to link to the slime link.
+ * Returns TRUE if the mind was successfully linked, FALSE otherwise.
+ */
+/datum/species/oozeling/stargazer/proc/link_mind(datum/mind/target_mind)
+	if(!validate_mind(target_mind, initial_connection = TRUE))
 		return FALSE
-	if(M in linked_mobs)
+	var/mob/living/carbon/human/owner = slimelink_owner.resolve()
+	if(target_mind in linked_minds)
 		return FALSE
 	if(!owner)
 		return FALSE
-	linked_mobs.Add(M)
-	to_chat(M, "<span class='notice'>You are now connected to [owner.real_name]'s Slime Link.</span>")
+	linked_minds.Add(target_mind)
 	var/datum/action/innate/linked_speech/action = new(src)
 	linked_actions.Add(action)
```

Comment:
You're looking for an action through assoc key `linked_actions[link_id]`, but this code only stores an action without assoc key.